### PR TITLE
Minor llama4 fixes

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

Fixes a couple nits/wrong defaults in Llama4 code for DynamicCache init - not sure we should even default to it, I think.
The scaling however has to be passed else `eager_attention_forward" will fail.

### CODE REVIEW

File: src/transformers/models/llama4/modeling_llama4.py [hunk 1]
```diff
@@ -534,10 +534,13 @@ def forward(
             inputs_embeds = self.embed_tokens(input_ids.to(self.embed_tokens.weight.device))
 
         if use_cache and past_key_values is None:
-            if self.config.get_text_config().get("attention_chunk_size") is not None:
+            if (
+                getattr(self.config, "attention_chunk_size", self.config.get_text_config().attention_chunk_size)
+                is not None
+            ):
```
Comment (ArthurZucker) [inline src/transformers/models/llama4/modeling_llama4.py:540]: not sure when the config would have it directly

File: src/transformers/models/llama4/modeling_llama4.py [hunk 2]
```diff
@@ -1117,7 +1120,7 @@ def forward(
             value_states,
             None,
             dropout=0.0 if not self.training else self.attention_dropout,
-            scaling=None,
+            scaling=self.scaling,
```
Comment (ArthurZucker) [inline src/transformers/models/llama4/modeling_llama4.py:1123]: super weird that this still work, can you run the test / check where this is from? (blame) might be expected 😐

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/llama4/modeling_llama4.py (hunk 2 above)
Comment (molbap) [reply src/transformers/models/llama4/modeling_llama4.py:1123]: changing tests, they were broken too :grinning:

Comment (molbap) [reply src/transformers/models/llama4/modeling_llama4.py:1123]: I ran the tests, it seems it works. I think it was broken during the interim after the recent VLMs refactor

Comment (molbap) [reply src/transformers/models/llama4/modeling_llama4.py]: ...but, because of TP, setting the scaling manually here is probably something that could break the vision attention module. I've set a `vision_eager_attention_forward` to at least fix the eager forward that is currently broken, while preserving the behaviour we had and err on the side of caution. @ArthurZucker alright with you?

Comment (vasqu) [reply src/transformers/models/llama4/modeling_llama4.py]: Pretty sure that sdpa, fa, and flex use the base `1/sqrt(head_dim)` when `scaling=None`. #37576 unintentionally changed the behaviour of eager by only using the passed scaling (before it manually calculated the scaling).

TP shouldn't affect the scaling tho, at least to my understanding. We slice by the total dim and only reduce by `num_heads`. 

For example, `num_heads=6, head_dim=64, TP=2` and given the view of `(bsz, seq_len, -1, head_dim)` --> this will result into shapes of `(bsz, seq_len, 3, 64)`. Iirc, old torch code also had times where they manually calculated the remaining heads based on the TP degree.

Comment (vasqu) [reply src/transformers/models/llama4/modeling_llama4.py]: Oh man, I see that we use the config's head dim and it's only set in the text attention :eyes: 

So in essence, the local `head_dim` was previously used in the vision attention. But depending on the TP degree, this is different. Did I understand it correctly? Imo, this is still a mistake, and not worth being BC: We essentially have different models based on different TP degrees - seems unreasonable to me and not the point of TP.

Comment (vasqu) [reply src/transformers/models/llama4/modeling_llama4.py]: Hmm, looking into the code https://github.com/zucchini-nlp/transformers/blob/b5757a4130ab297989c9fc074713db483b68ab55/src/transformers/models/llama4/modeling_llama4.py#L1212

It seems to me, the global `head_dim` was previously used before #37576 - not sure what the local `head_dim` would be.

Comment (molbap) [reply src/transformers/models/llama4/modeling_llama4.py]: I think `None` was passed (`self.scaling` was unset before the PR you linked). Definitely agree lol, the whole point of `tp_plan` is to have one model for all possible TP degrees. Passing `scaling=None` should be safe in all cases here, what was broken was mainly `eager_attention_forward` that did not compute `scaling` but instead used only the passed one, as you say... so now in `main` if you use `eager` you get a `float` * `NoneType` multiplication error here. 

So sdpa/flex/etc are fine IMO, just the eager path is currently broken, one simple way that should be `DTensor`-compatible when it comes to scaling could be e.g. `scaling = 1.0 / torch.sqrt(torch.tensor(query.size(-1), dtype=query.dtype, device=query.device))` WDYT?

Comment (vasqu) [reply src/transformers/models/llama4/modeling_llama4.py]: Shouldn't we be able to just pass `self.scaling` here @ https://github.com/huggingface/transformers/blob/34c1e29cdda6ba43dbe757c0707b73ad64efc5bb/src/transformers/models/llama4/modeling_llama4.py#L1120

But yea, if that doesn't work, then I'm definitely pro your solution :D

Comment (molbap) [reply src/transformers/models/llama4/modeling_llama4.py]: yeah that was my initial fix here https://github.com/huggingface/transformers/pull/38123/commits/9ffde0b97c136dd3e1d9a82c4e4ad9fb056473ee! but I wasn't sure of why things were in this state, it's good to put in writing :hugs:

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for digging! 
@vasqu happy if you upstream the changes to our interface to better support scaling / default etc to make it more user friendly and avoid us these kind of issues!

**Comment by vasqu** [comment]
Sounds good :D
